[quant][graphmode] Add add_relu pattern in skip values

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #33895 [quant][graphmode] insert prepack/unpack for aten::linear
* #33854 [quant][graphmode] Quantization pattern for aten::linear
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #33927 [quant][graphmode] Add Finalize function that inlines graph and produce quantized ops
* #33853 [quant][graphmode] Swap quantized functional linear with aten::linear
* #33481 [quant][graphmode] Handling ops doesn't require observation in insertObservers
* **#32816 [quant][graphmode] Add add_relu pattern in skip values**

Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208786](https://our.internmc.facebook.com/intern/diff/D20208786)

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 51e77330:


* **1/2** failures introduced in this PR
* **1/2** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=51e773305d3be7bc0baec3275be17d683838b008).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4377855) pytorch_linux_backward_compatibility_check_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93011057) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93011057))
<details>
<summary>
<code>Jan 30 03:24:45 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.</code>
</summary>

```
Jan 30 03:24:45 processing existing schema:  aten::values(Dict(str, t) self) -> (t[](*)) 
Jan 30 03:24:45 processing existing schema:  aten::values(Dict(int, t) self) -> (t[](*)) 
Jan 30 03:24:45 processing existing schema:  aten::values(Dict(float, t) self) -> (t[](*)) 
Jan 30 03:24:45 processing existing schema:  aten::values(Dict(Tensor, t) self) -> (t[](*)) 
Jan 30 03:24:45 processing existing schema:  aten::view(Tensor(a) self, int[] size) -> (Tensor(a)) 
Jan 30 03:24:45 processing existing schema:  aten::where(Tensor condition) -> (Tensor[]) 
Jan 30 03:24:45 processing existing schema:  aten::where.self(Tensor condition, Tensor self, Tensor other) -> (Tensor) 
Jan 30 03:24:45 processing existing schema:  aten::zeros_like(Tensor self, *, int? memory_format=None) -> (Tensor) 
Jan 30 03:24:45 processing existing schema:  aten::zeros_like.dtype(Tensor self, *, int dtype, int layout, Device device, bool pin_memory=False, int? memory_format=None) -> (Tensor) 
Jan 30 03:24:45 processing existing schema:  prim::Constant() -> (...) 
Jan 30 03:24:45 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.  
Jan 30 03:24:45  
Jan 30 03:24:45 Broken ops: [ 
Jan 30 03:24:45 	quantized::linear_relu_dynamic_fp16(Tensor X, Tensor W_prepack) -> (Tensor Y) 
Jan 30 03:24:45 ] 
Jan 30 03:24:45 + cleanup 
Jan 30 03:24:45 + retcode=1 
Jan 30 03:24:45 + set +x 
Jan 30 03:24:45 =================== sccache compilation log =================== 
Jan 30 03:24:45 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 30 03:24:45 Compile requests                 0 
```

</details>

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4377935) pytorch_linux_xenial_py3_clang5_asan_test (1/1)
**Step:** "Set Up CI Environment After attach_workspace" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=93010780) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=93010780)) :snowflake:
<details>
<summary>
<code>E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/xenial/pool/stable/amd64/docker-ce_18.09.4~3-0~ubuntu-xenial_amd64.deb  GnuTLS recv error (-54): Error in the pull function.</code>
</summary>

```
                                                                   82% [Working]               Get:26 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 intel-microcode amd64 3.20191115.1ubuntu0.16.04.2 [2,408 kB] 
82% [26 intel-microcode 20.5 kB/2,408 kB 1%] 84% [Working]                                                      18.7 MB/s 0s                                                                                 Get:27 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 amd64-microcode amd64 3.20191021.1+really3.20180524.1~ubuntu0.16.04.2 [30.8 kB] 
84% [27 amd64-microcode 23.5 kB/30.8 kB 76%]                       18.7 MB/s 0s 85% [Connecting to archive.ubuntu.com (91.189.88.162)]             18.7 MB/s 0s                                                                                 Get:28 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 linux-image-generic amd64 4.4.0.173.181 [2,486 B] 
85% [28 linux-image-generic 2,486 B/2,486 B 100%]                  18.7 MB/s 0s 85% [Working]                                                      18.7 MB/s 0s                                                                                 Get:29 http://archive.ubuntu.com/ubuntu xenial/universe amd64 moreutils amd64 0.57-1 [55.0 kB] 
85% [29 moreutils 12.6 kB/55.0 kB 23%]                             18.7 MB/s 0s 86% [Working]                                                      18.7 MB/s 0s                                                                                 Get:30 http://archive.ubuntu.com/ubuntu xenial/main amd64 tcl8.6 amd64 8.6.5+dfsg-2 [14.2 kB] 
86% [30 tcl8.6 4,871 B/14.2 kB 34%]                                18.7 MB/s 0s 86% [Working]                                                      18.7 MB/s 0s                                                                                 Get:31 http://archive.ubuntu.com/ubuntu xenial-updates/main amd64 thermald amd64 1.5-2ubuntu4 [187 kB] 
86% [31 thermald 4,869 B/187 kB 3%]                                18.7 MB/s 0s 87% [Working]                                                      18.7 MB/s 0s                                                                                 Get:32 http://archive.ubuntu.com/ubuntu xenial/main amd64 tk8.6 amd64 8.6.5-1 [12.2 kB] 
87% [32 tk8.6 7,447 B/12.2 kB 61%]                                 18.7 MB/s 0s 88% [Working]                                                      18.7 MB/s 0s                                                                                 Get:33 http://archive.ubuntu.com/ubuntu xenial/main amd64 xterm amd64 322-1ubuntu1 [607 kB] 
88% [33 xterm 4,869 B/607 kB 1%]                                   18.7 MB/s 0s 89% [Working]                                                      18.7 MB/s 0s                                                                                 Fetched 114 MB in 7s (15.6 MB/s) 
W: --force-yes is deprecated, use one of the options starting with --allow instead. 
E: Failed to fetch https://download.docker.com/linux/ubuntu/dists/xenial/pool/stable/amd64/docker-ce_18.09.4~3-0~ubuntu-xenial_amd64.deb  GnuTLS recv error (-54): Error in the pull function. 
 
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing? 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -449,6 +449,13 @@ graph(%self, %input, %inplace):
     %first_output = prim::CallMethod[name="forward"](%first_module, %input)
     %second_output = prim::CallFunction(%relu, %first_output, %inplace)
     return (%second_output) )");
+  // 2. Function - Module
```
What about `aten::add_` to `prim::CallFunction(relu)` (function-function)?

ZolotukhinM reviewed (approved):
Looks good modulo one comment.

dreiss reviewed (changes requested):
Can you add a test for functional relu?

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208786](https://our.internmc.facebook.com/intern/diff/D20208786)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208786](https://our.internmc.facebook.com/intern/diff/D20208786)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208786](https://our.internmc.facebook.com/intern/diff/D20208786)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add add_relu pattern in skip values"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20208786](https://our.internmc.facebook.com/intern/diff/D20208786)

[ghstack-poisoned]
